### ISSUE TITLE
How to access a Solidity custom error signature and it's arguments

### ISSUE THREAD

**Issue by bugbuzz** [ethereum/web3.py#2336]
* Version: web3==5.27.0
* Python: 3.9
* OS: osx

### What was wrong?

I have the following Solidity contract. The `exec()` function reverts with a [Solidity custom error](https://blog.soliditylang.org/2021/04/21/custom-errors/):

```solidity
error MyCustomError(address from, uint256 amount);

contract MyContract {
	function exec() external view {
		revert MyCustomError(msg.sender, 123);
	}
}
```

I want to access the custom error data anyhow after calling the `exec()` function, but I have only `web3.exceptions.ContractLogicError: execution reverted` exception with no arguments.

While the log shows it had reverted with the required details:

`
websockets.protocol       DEBUG    client < Frame(fin=True, opcode=<Opcode.TEXT: 1>, data=b'{"jsonrpc":"2.0","id":1,"error":{"code":3,"message":"execution reverted","data":"0xe13175d10000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000007b"}}\n', rsv1=False, rsv2=False, rsv3=False)
`

So I can't find any information how to access either the `"0xe13175d10000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000007b"` data or the decoded error arguments.

**Comment by kclowes** [comment]
Thanks for raising this! Short answer is no, there isn't really a good way to get the data. You can hack your way around it with some custom middleware (docs [here](https://web3py.readthedocs.io/en/stable/internals.html#middlewares)). The hacky workaround I came up with was to have my custom middleware make the request and then transform the response so you get the data key back. So something like: 

```
def transform_middleware(make_request, w3):
    def middleware(method, params):
        if method == 'eth_call':
            result = make_request(method, params)
            return {'result': {'data':  ...}}
        return make_request(method, params)
    return middleware
```

I'd like to find a nicer API for this. A few options I can think of off the top of my head are: 
1) Check for the data key in [here](https://github.com/ethereum/web3.py/blob/master/web3/_utils/method_formatters.py#L543) somewhere and still raise the error, but also return the data along with the message
1) Check for the data key [here-ish](https://github.com/ethereum/web3.py/blob/master/web3/manager.py#L163) and return it if it's there - not sure how useful this would be
1) Have a convenience method to get the raw result back with no processing. Something like: `w3.provider.get_raw_response('method_name')`
1) Pass a flag to any of the methods that raise the ContractLogicError that just returns the raw response if an error comes back

These aren't very well thought through, just tossing some ideas out there :)

**Comment by fselmo** [comment]
closed by #2795 and #2922
